lib/watcherjob/watcherjob.go

```diff
@@ -0,0 +1,230 @@
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport-plugins/lib"
+	"github.com/gravitational/teleport-plugins/lib/logger"
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+
+	"google.golang.org/grpc"
+)
+
+const DefaultQueueSize = 32
+const DefaultEventFuncTimeout = time.Second * 5
+
+type EventFunc func(context.Context, types.Event) error
+
+type Config struct {
+	Watch            types.Watch
+	MaxConcurrency   int
+	QueueSize        int
+	EventFuncTimeout time.Duration
+}
+
+type job struct {
+	lib.ServiceJob
+	events    types.Events
+	config    Config
+	eventFunc EventFunc
+	queues    map[eventQueueKey]*eventQueue
+	queueMu   sync.Mutex
+	queueCond *sync.Cond
+}
+
+type eventQueueKey struct {
+	kind string
+	name string
+}
+
+type eventQueue struct {
+	key  eventQueueKey
+	ch   chan types.Event
+	busy int
+}
+
+func NewJob(client *client.Client, config Config, fn EventFunc) lib.ServiceJob {
+	client = client.WithCallOptions(grpc.WaitForReady(true)) // Enable backoff on reconnecting.
+	return NewJobWithEvents(client, config, fn)
+}
+
+func NewJobWithEvents(events types.Events, config Config, fn EventFunc) lib.ServiceJob {
+	if config.MaxConcurrency == 0 {
+		config.MaxConcurrency = runtime.GOMAXPROCS(0)
+	}
+	if config.QueueSize == 0 {
+		config.QueueSize = DefaultQueueSize
+	}
+	if config.EventFuncTimeout == 0 {
+		config.EventFuncTimeout = DefaultEventFuncTimeout
+	}
+	job := job{
+		events:    events,
+		config:    config,
+		eventFunc: fn,
+		queues:    make(map[eventQueueKey]*eventQueue),
+	}
+	job.queueCond = sync.NewCond(&job.queueMu)
+	job.ServiceJob = lib.NewServiceJob(func(ctx context.Context) error {
+		ctx, cancel := context.WithCancel(ctx)
+		log := logger.Get(ctx)
+
+		lib.MustGetProcess(ctx).OnTerminate(func(_ context.Context) error {
+			cancel()
+			job.queueCond.Broadcast()
+			return nil
+		})
+
+		for {
+			err := job.eventLoop(ctx)
+			switch {
+			case trace.IsConnectionProblem(err):
+				log.WithError(err).Error("Failed to connect to Teleport Auth server. Reconnecting...")
+			case trace.IsEOF(err):
+				log.WithError(err).Error("Watcher stream closed. Reconnecting...")
+			case lib.IsCanceled(err):
+				log.Debug("Watcher context is cancelled")
+				// Context cancellation is not an error
+				return nil
+			default:
+				log.WithError(err).Error("Watcher event loop failed")
+				return trace.Wrap(err)
+			}
+		}
+	})
+	return &job
+}
+
+// eventLoop reads the events from watcher and spawns temporary micro-queues.
+func (job *job) eventLoop(ctx context.Context) error {
+	watcher, err := job.events.NewWatcher(ctx, job.config.Watch)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := watcher.Close(); err != nil {
+			logger.Get(ctx).WithError(err).Error("Failed to close a watcher")
+		}
+	}()
+
+	if err := job.waitInit(ctx, watcher, 5*time.Second); err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
Nit: 5 seconds is a bit short.  Init events aren't always sent immediately.  E.g. if the teleport instance is using a cache and the watcher is created during a cache reset, then the init event is not sent until after the cache reset is complete.  Something like a 30 second timeout is probably better.

Reply:
@fspmarshall 

If we haven't get `OpInit` we don't proceed further to process the events.

Is it event possible to get `OpPut`, `OpPut`, ... `OpInit` event sequence? Then do we really need to wait for `OpInit` at all?

Reply:
Sorry, my wording was ambiguous.  When I say that `OpInit` isn't always sent immediately, I don't mean that it might arrive after other events.  It is always the first event.  I just mean that sometimes there is an extra delay between the time when the watcher is created, and the time when `OpInit` is sent.

In theory, it would be possible to go straight to the event processing loop, and simply ignore the `OpInit` event when it comes through.  Why we don't typically do this is because teleport instances which are unhealthy will queue up watchers, and then sent the `OpInit` event when the instance become healthy.  By enforcing a timeout on how long we're willing to wait for `OpInit`, we're limiting the amount of time that we wait for the instance to become healthy.  Since clients typically interact with teleport via a load balancer, having a reasonable timeout for `OpInit` allows to to attempt to "switch" to a healthy instance if the one we are currently connected to is taking too long to become healthy.